server/src/main/java/org/elasticsearch/search/runtime/LongScriptFieldRangeQuery.java

```diff
@@ -21,15 +24,26 @@
 
     public LongScriptFieldRangeQuery(
         Script script,
-        Function<LeafReaderContext, AbstractLongFieldScript> leafFactory,
         String fieldName,
+        Query approximation,
+        Function<LeafReaderContext, AbstractLongFieldScript> leafFactory,
         long lowerValue,
         long upperValue
     ) {
-        super(script, leafFactory, fieldName);
+        super(script, fieldName, approximation, leafFactory);
         this.lowerValue = lowerValue;
         this.upperValue = upperValue;
-        assert lowerValue <= upperValue;
+        assert lowerValue <= upperValue : lowerValue + " <= " + upperValue;
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        Query newApprox = approximation().rewrite(reader);
+        if (newApprox == approximation()) {
+            return this;
+        }
+        // TODO move rewrite up
+        return new LongScriptFieldRangeQuery(script(), fieldName(), newApprox, scriptContextFunction(), lowerValue, upperValue);
```

Comment:
I think it'd be cleaner if `rewrite` were final on `AbstractScriptFieldQuery` and subclasses had to implement a `rebuildWithRewrittenApproximation` but I thought that'd require touching all subclasses and wanted to save that for a followup. I suppose we could make `rebuildWithRewrittenApproximation` throw an exception by default and just override it where we need it....